alacritty_terminal/src/term/mod.rs

```diff
@@ -2118,10 +2117,23 @@ impl ansi::Handler for Term {
     }
 
     #[inline]
-    fn set_scrolling_region(&mut self, region: Range<Line>) {
-        trace!("Setting scrolling region: {:?}", region);
-        self.scroll_region.start = min(region.start, self.grid.num_lines());
-        self.scroll_region.end = min(region.end, self.grid.num_lines());
+    fn set_scrolling_region(&mut self, top: usize, bottom: usize) {
+        if top >= bottom {
+            debug!("Invalid scroll region: ({};{})", top, bottom);
+            return;
+        }
+
+        // Bottom should be included in the range, but range end is not
+        // usually included. One option would be to use an inclusive
+        // range, but instead we just let the open range end be 1
+        // higher.
+        let start = Line(top - 1);
+        let end = Line(bottom);
```

Comment:
I'm going to keep making some noise about these pervasive `Line` constructs. :boom: 

We're moving away from a `Range<Line>`, which I can understand since ranges are less clear in this context than "top" and "bottom", but we're replacing it with two `usize` values. This seems most natural to me, but defeats the point of the `Line`. I think this is further evidence that the `Line` in general is doing more harm than good in it's current form, and would be better off eliminated.

That said, we use `Line` in many places as indices, and this isn't necessarily bad, however, we seem to be using `Line` to mean both index, and count, depending on context, and sometimes just a plain old `usize` (re, this PR)... This is further confused when interfacing with specs (we don't control) that use 1 indexing.

https://github.com/jwilm/alacritty/blob/c1f089970fd3d4b9137d07647f8cd028b2b5b3a9/alacritty_terminal/src/grid/mod.rs#L142-L150

https://github.com/jwilm/alacritty/blob/c1f089970fd3d4b9137d07647f8cd028b2b5b3a9/alacritty_terminal/src/grid/mod.rs#L286

Either we make a type for each kind of thing, i.e. `Line`, `Lines`, `OneIndexedLine`, or we should bail on the concept altogether IMO.

I've made this argument before, but I'm making it again. Sorry.

Reply:
To go further off the deep end on this topic, let's talk about indices and counts in general here. Please forgive my rambling:

```rust
// This is an *empty* array `[]`.
let empty = [];
let n = 0;
let i = None;

// This is an array with a 0th element `["A"]`.
let single = ["A"];
let n = 1;
let i_a = Some(0);
let i_wtf = Some(1);

// You can see where this is going...
let many = ["A", "B", "C"];
let n = 3;
let i_c = Some(2);
let i_wtf = Some(3);

// So we can see two important facts about indices with respect to counts:
//
// 1. When the count is 0, there isn't an index to be had at all
// 2. When the count is > 0, the index is bound to be < the count
//
// OK.

// So we need to decide if we want a type for cases when we already have an
// element in mind, so no optional is needed, or if we want to store an index
// generally along side some collection, making an optional needed.

// For Alacritty's purposes, we generally already have elements in mind, and
// simply want to address them. Maybe I'm just not giving enough credit to the
// case when the grid is empty though.
//
// For the next bit tho, let's pretend like we've always got at least one
// element...
struct Count(usize);
struct Index(usize);

assert_eq!(many.count(), Count(3));
assert_eq!(many.index_of("A"), Index(0));

// Indices shouldn't be added to each other, but instead with a count.
assert_eq!(Index(0) + Index(10), ???);
assert_eq!(Index(0) + Count(10), Index(10));

// This doesn't help with the fundamental issue motivating Alacritty's index
// types in the first place however, which is mainly to prevent confusing
// columns and lines.
//
// Here we call a poorly defined function, since it's *VERY* easy to call in
// the wrong order.
let column = 0;
let line = 1;
poorly_designed_function(column, line);

// If Rust had keyword arguments we could have designed a function:
properly_designed_function(column: column, line: line);
// And struct style elision might even allow us to write
properly_designed_function(column, line);
// As in the poorly designed function before... (this syntax is contentious)

// However in the real world (for now), we can write
let column = Column(0);
let line = Line(1);
our_designed_function(column, line);

// This comes with other benefits. Like causing a type error when I fail to
// copy and paste this simple arithmetic. This is a somewhat contrived example,
// but does honestly prevent some errors.
let ca = Column(0);
let cb = Column(1);
let la = Line(0);
let lb = Line(3);
let cr = ca + cb;
let lr = la + cb;

// But, here's where the problem begins! Now we're tempted to do this
let current_position = Index(9);
let current_lines = current_position.0 + 1;
// Which gets more confusing if we mistakenly use an index as a count!
let lines = Index(10);
let position = lines.0 - 1;
// And we forget to write `.0` and end up with indices where counts should be!

// But this is just a matter of `+ 1` vs `- 1` you say! Yes, and the other
// issue was just a matter of `Column` vs `Line`... (Un)fortunately, these
// things make a difference.

// Of course none of this talks about translating coordinate systems that grow
// from the top, or bottom.

// Is this the top or the bottom?
Index(0)

// Luckily, this is why god gave us names.
let bottom = Index(0);

// We could encode this into the type as well.
let index = BottomIndexedIndex(0);
// But I can't think of a pretty way to do this...
let index = bottom::Index(0);
// Better, but still...

// So we need (Line, Count + Index) and (Column, Count + Index).
struct Count<T>(usize, PhantomData<T>);
struct Index<T>(usize, PhantomData<T>);
struct Line;
struct Column;

// Now between the value types (Count, and Line), the marker types (Line, and
// Column), and our binding names (top, and bottom), we've completely specified
// the three critical aspects of our grid's elements. Although, in some cases
// the binding name will imply the type, this can't be relied on in general.
// This is however a step up in expressive power than we have today.
let top = Index::<Line>::new(23);
let replacement = Index::<Column>::new(13);
let visible = Count::<Line>::new(1337);

// However this method requires a `new` function, and suddenly I find myself
// wondering about the overhead of these supposedly helpful types.

// Finally I arrive at
struct Line(usize);   // An index.
struct Lines(usize);  // A count.

let top = Line(23);
let history = Lines(1337);
assert_eq!(top + history, Line(1360));

// Simple, elegant. DHH would be proud.

// I'm sure there's more to this still, but I'm going to sleep now.
```

I'm worried the most *interesting* point in all this remains unaddressed, which is that an `Index` is generally an `Option<usize>`, which isn't something I think we care to talk about much.